extmod/machine_pulse.c

```diff
@@ -30,20 +30,35 @@
 
 #if MICROPY_PY_MACHINE_PULSE
 
-MP_WEAK mp_uint_t machine_time_pulse_us(mp_hal_pin_obj_t pin, int pulse_level, mp_uint_t timeout_us) {
+mp_uint_t machine_time_pulse_us(mp_hal_pin_obj_t pin, int pulse_level, mp_uint_t timeout_us) {
+    mp_uint_t nchanges = 2;
     mp_uint_t start = mp_hal_ticks_us();
-    while (mp_hal_pin_read(pin) != pulse_level) {
-        if ((mp_uint_t)(mp_hal_ticks_us() - start) >= timeout_us) {
-            return (mp_uint_t)-2;
-        }
-    }
-    start = mp_hal_ticks_us();
-    while (mp_hal_pin_read(pin) == pulse_level) {
-        if ((mp_uint_t)(mp_hal_ticks_us() - start) >= timeout_us) {
-            return (mp_uint_t)-1;
+    for (;;) {
+        // Sample ticks and pin as close together as possible, and always in the same
+        // order each time around the loop.  This gives the most accurate measurement.
+        mp_uint_t t = mp_hal_ticks_us();
+        int pin_value = mp_hal_pin_read(pin);
+
+        if (pin_value == pulse_level) {
+            // Pin is at desired value.  Flip desired value and see if we are done.
+            pulse_level = 1 - pulse_level;
+            if (--nchanges == 0) {
+                return t - start;
```

Comment:
Shouldn't that be `--nchanges <= 0`. Otherwise the mechanism of waiting for the opposite pulse by setting nchanges to 3 would not work for all start levels. I know that  nchanges was set hard to 2 here, but you discussed making that an optional argument.